### PR TITLE
added tasks related to "operations_in_a_natural_way"

### PR DESCRIPTION
**Pull request by Ypostolov**
These tasks relate to the topic  "structures and classes"

### CODE REVIEW

File: week_4/class_rational_part_1/class_rational_part_1.cpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  Rational() {
+    p = 0;
+    q = 1;
+    divider = 1;
+  }
+
+  Rational(int numerator, int denominator) {
+    if (numerator == 0) {
+      p = numerator;
+      q = 1;
```
Comment (ZhdanovP) [inline week_4/class_rational_part_1/class_rational_part_1.cpp:14]: @Ypostolov let's implement copy constructor for that

File: week_4/class_rational_part_1/class_rational_part_1.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+ private:
+  int divider;
+  int p;
+  int q;
+};
+
+int main() {
+  {
```
Comment (ZhdanovP) [inline week_4/class_rational_part_1/class_rational_part_1.cpp]: @Ypostolov Why do we need this extra scope?

File: week_4/class_rational_part_1/class_rational_part_1.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    const Rational r(3, 10);
+    if (r.Numerator() != 3 || r.Denominator() != 10) {
+      std::cout << "Rational(3, 10) != 3/10" << std::endl;
+      return 1;
+    }
+  }
+  {
+    const Rational r(8, 12);
+    if (r.Numerator() != 2 || r.Denominator() != 3) {
+      std::cout << "Rational(8, 12) != 2/3" << std::endl;
```
Comment (ZhdanovP) [inline week_4/class_rational_part_1/class_rational_part_1.cpp]: @Ypostolov Divide this test for separate fucntions

**Review by ZhdanovP — CHANGES_REQUESTED** [review]

File: week_4/class_rational_part_1/class_rational_part_1.cpp (hunk 1 above)
Comment (Ypostolov) [reply week_4/class_rational_part_1/class_rational_part_1.cpp:14]: Emplemented.
Please check 6f94025d2712317c65697f93f5dc11e2eb1b0da5

File: week_4/class_rational_part_1/class_rational_part_1.cpp (hunk 2 above)
Comment (Ypostolov) [reply week_4/class_rational_part_1/class_rational_part_1.cpp]: We use extra scope as access modifiers,according to the template from the task statement, we use the same name for the class instance when creating it.

File: week_4/class_rational_part_1/class_rational_part_1.cpp (hunk 3 above)
Comment (Ypostolov) [reply week_4/class_rational_part_1/class_rational_part_1.cpp]: Please clarify the comment

File: week_4/class_rational_part_1/class_rational_part_1.cpp (hunk 1 above)
Comment (ZhdanovP) [reply week_4/class_rational_part_1/class_rational_part_1.cpp:14]: @Ypostolov I don't see copy constructor in 6f94025d2712317c65697f93f5dc11e2eb1b0da5
And please, do not squash commits during review which you are referencing

File: week_4/class_rational_part_1/class_rational_part_1.cpp (hunk 2 above)
Comment (ZhdanovP) [reply week_4/class_rational_part_1/class_rational_part_1.cpp]: @Ypostolov According to your explanation, the main reason of this scope is using the same name for the object. Am I right?

File: week_4/class_rational_part_1/class_rational_part_1.cpp (hunk 3 above)
Comment (ZhdanovP) [reply week_4/class_rational_part_1/class_rational_part_1.cpp]: @Ypostolov Do not use extra scopes in main function create few functions for each test instead and call them sequentially from the main
...
test_1();
test_2();
...

File: week_4/class_rational_part_1/class_rational_part_1.cpp (hunk 1 above)
Comment (ZhdanovP) [reply week_4/class_rational_part_1/class_rational_part_1.cpp:14]: @Ypostolov I see that changes in different commit 09c8a4ad275340d60d9684d1a5ac510accdfe17c
So, why this copy constructor do not copy anything inside?

Comment (Ypostolov) [reply week_4/class_rational_part_1/class_rational_part_1.cpp:14]: Fixed in 682be2784f7a4b735014298dac5a1905393ae59a
Please take a look.

File: week_4/class_rational_part_1/class_rational_part_1.cpp (hunk 3 above)
Comment (Ypostolov) [reply week_4/class_rational_part_1/class_rational_part_1.cpp]: Fixed in 682be2784f7a4b735014298dac5a1905393ae59a
Please take a look.

File: week_4/class_rational_part_1/class_rational_part_1.cpp (hunk 2 above)
Comment (Ypostolov) [reply week_4/class_rational_part_1/class_rational_part_1.cpp]: Yes, because from scopes depend lifetime of variables.

**Review by ZhdanovP — APPROVED** [review]

**Pull request merged**